### ISSUE TITLE
Nested collection filters apply to different items

### ISSUE THREAD

**Issue by davla** [ravendb/ravendb#18613]
When having two levels of nested collections, it looks like conditions don't work as expected.

In particular, a filter like this:

```cs
await session.Query<EntityWithNestedCollection>()
            .Where(x => x.LevelOne.Any(y => y.Name == nameWithoutCool && y.LevelTwo.Any(z => z.IsCool)))
```

behaves somewhat like this one:

```cs
await session.Query<EntityWithNestedCollection>()
    .Where(
        x => x.LevelOne.Any(y => y.Name == nameWithoutCool) &&
             x.LevelOne.Any(y => y.LevelTwo.Any(z => z.IsCool))
    )
```

That is, the `&&` is disjoint over two different items in the first-level collection.

For a bit more context, this is a test that fails.

```cs
[TestClass]
public class BugReportTests : RavenDbTestBase
{
    [TestMethod]
    public async Task AndInNestedCollectionCondition_OperatesOnSameNestedCollectionItem()
    {
        // arrange
        const string nameWithCool = "cool";
        const string nameWithoutCool = "not-cool";
        var entity = new EntityWithNestedCollection
        {
            LevelOne =
            [
                new EntityWithCollection()
                {
                    Name = nameWithCool,
                    LevelTwo = [new CoolEntity() { IsCool = true }]
                },
                new EntityWithCollection()
                {
                    Name = nameWithoutCool,
                    LevelTwo = [new CoolEntity() { IsCool = false }]
                }
            ]
        };

        using var session = Store.OpenAsyncSession();
        await session.StoreAsync(entity);
        await session.SaveChangesAsync();

        // act
        var coolEntitiesInWithoutCool = await session.Query<EntityWithNestedCollection>()
            .Where(x => x.LevelOne.Any(y => y.Name == nameWithoutCool && y.LevelTwo.Any(z => z.IsCool)))
            .ToListAsync();

        // assert - this fails because `entity` is returned from the query
        coolEntitiesInWithoutCool.Should()
            .BeEmpty();
    }

    public class EntityWithNestedCollection : BaseStorageEntity
    {
        public List<EntityWithCollection> LevelOne { get; set; } = [];
    }

    public class EntityWithCollection
    {
        public required string Name { get; set; }

        public List<CoolEntity> LevelTwo { get; set; } = [];
    }

    public class CoolEntity
    {
        public bool IsCool { get; set; }
    }
}
```

**Comment by ayende** [comment]
Yes, that is the expected behavior, I'm afraid.
See here for deeper discussion: 

https://ravendb.net/docs/article-page/6.0/csharp/indexes/indexing-nested-data

Basically, you need to either do a fanout or intersection

**Comment by davla** [comment]
Thank you for the quick reply!

Good to know that it's an addressed use-case with good documentation. We actually ended up implementing the "Simple index" approach, but now I can see we have more options we were not aware of. Thank you for pointing us in the right direction.

**Comment by loraderon** [comment]
> Yes, that is the expected behavior, I'm afraid. 

This should really be caught at compile time (roslyn analyzer) or at least at runtime (linq provider exception).

We are in the process of migrating to raven and sometimes it's just not obvious that queries will "work" but produce the wrong results. How can we as "new" users be aware of these expected behaviors?

Can this query be satisfied with DocumentQuery or RQL?

```cs
private record Person(Pants[] Pants);
private record Pants(string Color, string Size);

[RavenFact(RavenTestCategory.ClientApi)]
public void AnyWithAnd()
{
    using var store = GetDocumentStore();

    using (var session = store.OpenSession())
    {
        session.Store(new Person([
            new("Red", "M" ),
        ]));
        session.Store(new Person([
            new("Red", "L" ),
            new("Blue", "M" ),
        ]));

        session.SaveChanges();
    }

    using (var session = store.OpenSession())
    {
        var personsWithRedMediumPants = session.Query<Person>()
            .Where(x => x.Pants.Any(p => p.Color == "Red" && p.Size == "M"))
            .ToList();

        Assert.Single(personsWithRedMediumPants);
    }            
}
```

**Comment by loraderon** [comment]
Looks like there is some action here https://issues.hibernatingrhinos.com/issue/RavenDB-21777

**Comment by Lwiel** [comment]
We have [an issue](https://issues.hibernatingrhinos.com/issue/RavenDB-21777/Wrong-RQL-generated-when-using-LINQ-.Any-that-contains-logical-AND-in-arrays) created for v6.1. We'd like to throw an exception for such queries on auto map index.

In order to get the expected results you should create a static index that will create index entries containing both `Color` and `Size`. We have [another issue](https://issues.hibernatingrhinos.com/issue/RavenDB-21773/Handling-of-fanout-auto-map-indexes) for supporting fanout auto map indexes, which is the subject of discussion for the future.

**Comment by loraderon** [comment]
@Lwiel Great, I think that throwing exceptions for unsupported queries is helpful.

It would be even better with Rolsyn analyzers as well https://github.com/ravendb/ravendb/discussions/17089.

**Comment by Lwiel** [comment]
[PR with exception throwing](https://github.com/ravendb/ravendb/pull/19052) got merged for v6.1
